### ISSUE TITLE
Promise.map doesn't await all promises

### ISSUE THREAD

**Issue by benbuckman** [petkaantonov/bluebird#1198]
1) What version of bluebird is the issue happening on?

v3.4.1, v2.9.24

2) What platform and version? (For example Node.js 0.12 or Google Chrome 32)

Node v4.4.2

3) Did this issue happen with earlier version of bluebird?

Yes

The [docs for `Promise.map`](http://bluebirdjs.com/docs/api/promise.map.html) state (emphasis mine),

> **Promises returned by the mapper function are awaited for** and the returned promise doesn't fulfill until all mapped promises have fulfilled as well.

I understand the first part to mean that all the promises are awaited on _even if one of them fails_.

So for example, this works as expected:

```js
'use strict';
const Promise = require('bluebird');

// Fails when `n === 10`.
function process(n) {
  if (n === 5) {
    return Promise.reject(new Error(n + " is bad"));
  } else {
    return Promise.resolve(n);
  }
}

// Log and passthru resolution.
function wrappedFn(n) {
  return process(n).then((result) => {
    console.log(n, 'ok', result);
    return result;
  }).catch((err) => {
    console.log(n, 'failed', err.message);
    throw err;
  });
}

const nums = [];
for (let i = 1; i <= 10; i++) nums.push(i)

Promise.map(nums, (n) => {
  return wrappedFn(n);
})
// I expect this _not_ to get hit b/c of the failure with n==5.
.then(() => {
  console.log('success');
})
// I expect this to be called after ALL the `map` iterations.
.catch((err) => {
  console.log('failure:', err.message);
})
.finally(() => {
  console.log('done');
});
```

It outputs:

```
1 'ok' 1
2 'ok' 2
3 'ok' 3
4 'ok' 4
6 'ok' 6
7 'ok' 7
8 'ok' 8
9 'ok' 9
10 'ok' 10
5 'failed' '5 is bad'
failure: 5 is bad
done
```

Which is what I would expect.

But if there is a _decreasing delay_ added, then it breaks:
```js
'use strict';
const Promise = require('bluebird');

// Fails when `n === 5`.
function process(n) {
  // Decreasing delay with each iteration
  const z = 300 - (n * 10);
  return Promise.delay(z).then(() => {
    if (n === 5) {
      throw new Error(n + " is bad");
    } else {
      return n;
    }
  });
}

// Log and passthru resolution.
function wrappedFn(n) {
  return process(n).then((result) => {
    console.log(n, 'ok', result);
    return result;
  }).catch((err) => {
    console.log(n, 'failed', err.message);
    throw err;
  });
}

const nums = [];
for (let i = 1; i <= 13; i++) nums.push(i)

Promise.map(nums, (n) => {
  return wrappedFn(n);
})
// I expect this _not_ to get hit b/c of the failure with n==5.
.then(() => {
  console.log('success');
})
// I expect this to be called after ALL the `map` iterations.
.catch((err) => {
  console.log('failure', err.message);
})
.finally(() => {
  console.log('done');
});
```

That outputs:

```
13 'ok' 13
12 'ok' 12
11 'ok' 11
10 'ok' 10
9 'ok' 9
8 'ok' 8
7 'ok' 7
6 'ok' 6
5 'failed' '5 is bad'
failure 5 is bad
done
4 'ok' 4
3 'ok' 3
2 'ok' 2
1 'ok' 1
```

Basically, with this decreasing delay on each iteration, it appears to not be waiting on all the promises anymore. (It bails out of the `map` before all the iterations have finished.)

Is this intentional or a bug? And why is it different from the first example?
(In the 2nd example, if the `delay` interval changes to `const z = 300;` – the same on every iteration instead of decreasing – then it seems to work again.)

I'm pretty confused.

Thanks!

**Comment by petkaantonov** [comment]
The next sentence after the one you quoted says what happens when any promise is rejected 

>  If any promise in the array is rejected, or any promise returned by the mapper function is rejected, the returned promise is rejected as well.

**Comment by benbuckman** [comment]
Then why, in the first code example, does it finish all 10 before it reports the failure from 5? It seems like the behavior is inconsistent.

**Comment by spion** [comment]
Because promise handlers are not executed immediately, instead they're guaranteed to execute after all synchronous code in the current user stack completes executing. Turns out at that point, all the promises are already rejected in the first example.
